xcp/accessor.py

```diff
@@ -332,9 +337,11 @@ def openAddress(self, address):
         url = urllib.parse.unquote(address)
 
         self.ftp.voidcmd('TYPE I')
-        s = self.ftp.transfercmd('RETR ' + url).makefile('rb')
+        # Collect openened sockets and close them in self.finish() when we shut down:
+        socket = self.ftp.transfercmd('RETR ' + url)
+        self.sockets.append(socket)
         self.cleanup = True
-        return s
+        return socket.makefile('rb')
```

Comment:
It shouldn't be necessary to keep a list of sockets around since they are reference counted. From the docs of socket.close:

```
Mark the socket closed. The underlying system resource (e.g. a file descriptor) is also closed when all file objects from [makefile()](https://docs.python.org/3/library/socket.html#socket.socket.makefile) are closed. Once that happens, all future operations on the socket object will fail.
```

So something like this should suffice:

```
f = socket.makefile('rb')
socket.close()
return f
```

Reply:
I checked the cpython code and as far as I can see, closing the socket would not affect the created file object and the following test confirmed that:

```
$ cat stest.py
import socket
import threading
import time

def write_thread(s):
    time.sleep(1)
    s.send(b'foo')
    time.sleep(1)
    s.close()

def get_file():
    writesock, sock = socket.socketpair()

    f = sock.makefile('rb')
    sock.close()

    threading.Thread(target=write_thread, args=(writesock,)).start()

    return f

f = get_file()
print(f.read())

$ python3 --version
Python 3.6.8

$ python3 stest.py
b'foo'

$ python2 --version
Python 2.7.18

$ python2 stest.py
foo
```

Reply:
@rosslagerwall I stand corrected! I've no idea what went wrong in my earlier attempts, CI has passed, looks fine.